src/stirling/bpf_tools/task_struct_resolver.cc

```diff
@@ -137,8 +137,54 @@ StatusOr<TaskStructOffsets> ScanBufferForFields(const struct buf& buf,
 
   TaskStructOffsetsManager offsets_manager(&task_struct_offsets);
 
+#if AARCH64
+  // On aarch64, the cpu_context struct inside the thread_struct, can contain the same address as
+  // the group_leader value of the task_struct. Since the thread_struct is always at the end of the
+  // task_struct, we can just stop searching when we get to the thread_struct. To identify the
+  // offset of the thread_struct, we search for the address of thread local storage which is always
+  // stored `sizeof(cpu_context) + some padding` after the beginning of the thread struct.
+  uint64_t tls_addr;
+  // ARM magic to get the system register TPIDR_EL0, which stores the address of thread local
+  // storage.
+  asm volatile("mrs %0, TPIDR_EL0" : "=r"(tls_addr));
+  VLOG(1) << "TLS address = " << tls_addr;
+  // For now we iterate through the buffer twice. At some point we should refactor offsets_manager
+  // to eliminate the need to iterate through the buffer twice.
+  int stop_at_offset = -1;
```

Comment:
(note: it won't let me attach this comment to line 154 on the "right hand side" of the diff).

Related, but, it seems that the buffer index must be at least 14 and the byte offset 112 to find the tls addr. We could re-write line 154 like so:
```
for (const auto& [idx, val] : Enumerate(buf.u64words + tp_value_offset_in_thread_struct / sizeof(uint64_t))) { ...
```
I'm not sure if that compiles, but there might be some way to start the iteration at a crafted location using some arithmetic like above.

Reply:
What you suggested doesn't quite work because of the Enumerate, but I did something to this effect.